text/0003-execution-environment.md

```diff
@@ -0,0 +1,102 @@
+# Execution Environments
+
+## Summary
+
+This RFC introduces settings for controlling what execution environment (Node.js, Bun, Deno) will be used for a package during:
+
+* runnings its lifecycle scripts
```

Comment:
One thing to note is that lifecycle scripts can technically choose to execute `node`, `deno`, and `bun` in the same command because they are shell scripts.

Reply:
There should still be a primary runtime for installing CLI app.

Lifecycles OTOH can take advantage of `executionEnv.nodeVersion`, `executionEnv.denoVersion`, `executionEnv.bunVersion`.

I don't quite understand the `building` item, is it lifecycle?

Reply:
The building part is when the package is installed as a dependency. If it has a "postinstall" script, it will be executed to build the package. Or if it has a `binding.gyp` file, then `node-gyp` will run to build the package (it can still run node under the hood).

Reply:
> Lifecycles OTOH can take advantage of executionEnv.nodeVersion, executionEnv.denoVersion, executionEnv.bunVersion.

ok, so you suggest to keep the `nodeVersion` field, add [runtime]Version fields and a `jsRuntime` field. In that case, I guess jsRuntime will always be used, when the package is installed as a dependency (so the localOnly field is not needed).

Reply:
I see. So it can take advantage of `node`, `deno`, and `bun` from PATH.

I have this suggestion: Instead of `jsRuntime`, we should have `nodeVersion`, `denoVersion`, `bunVersion`, and optional `cliRuntime`. By default, defining only 1 `{x}Version` would set `cliRuntime` to `{x}`. If more than 1 `{x}Version` is defined, `cliRuntime` is required. What do you think?

Reply:
For reference, there is also this RFC: https://github.com/openjs-foundation/package-metadata-interoperability-collab-space/issues/15

Reply:
In the referenced issue they have also pointed out that there is an existing field for specifying runtime environments: engine.

Reply:
I am fine with your suggestion to have `nodeVersion`, `denoVersion`, `bunVersion`. I am not sure about the rest of the suggestion though. Especially as having a cliRuntime should be optional, so automatically generating it doesn't makes sense.

Something like an object with setting could work too:

```json
{
  "pnpm": {
    "executionEnv": {
      "nodeRuntime": {
        "version": "20.16.0",
        "cli": true
      }
    }
  }
}
```

Reply:
> I am fine with your suggestion to have `nodeVersion`, `denoVersion`, `bunVersion`. I am not sure about the rest of the suggestion though. Especially as having a cliRuntime should be optional, so automatically generating it doesn't makes sense.
> 
> Something like an object with setting could work too:
> 
> ```json
> {
>   "pnpm": {
>     "executionEnv": {
>       "nodeRuntime": {
>         "version": "20.16.0",
>         "cli": true
>       }
>     }
>   }
> }
> ```

I'm not sure about adding another nesting level. Besides that, this also creates an invalid state where `nodeRuntime.cli`, `denoRuntime.cli`, and `bunRuntime.cli` are all defined, compared to `cliRuntime` which doesn't have invalid state.

> Especially as having a cliRuntime should be optional, so automatically generating it doesn't makes sense.

We can improve it a bit: `cliRuntime` is only required when the package define `bin` and there are more than 1 `{x}Version`.

Reply:
Right, but even when there is a `bin` field, this is optional. The nesting can be not required. Like in the dependencies in `Cargo.toml`. For instance:

```json
{
  "pnpm": {
    "executionEnv": {
      "runtime": {
        "deno": "1.1.0",
        "node": {
          "version": "20.16.0",
          "cli": true
        }
    }
  }
}
```